synapse/storage/databases/main/user_directory.py

```diff
@@ -511,7 +510,6 @@ def _update_profile_in_user_dir_txn(txn: LoggingTransaction) -> None:
                     table="user_directory_search",
                     keyvalues={"user_id": user_id},
                     values={"value": value},
-                    lock=False,  # We're only inserter
```

Comment:
Unique index exists here for postgres: https://github.com/matrix-org/synapse/blob/25f43faa70f7cc58493b636c2702ae63395779dc/synapse/storage/schema/main/delta/42/user_dir.py#L59

... but not for sqlite. It's marked as unsafe to upsert into on sqlite here: https://github.com/matrix-org/synapse/blob/c3a4780080a5bcb04132283c0f32f7452655792a/synapse/storage/database.py#L534-L538

Does that mean that with this change we're now going to start locking `user_directory_search` on sqlite deployments? Ahh, but locking a table is a no-op on sqlite, presumably because there's at most one worker writing? https://github.com/matrix-org/synapse/blob/8c94dd3a277d4e11192f98a9ca32cb6638606b66/synapse/storage/engines/sqlite.py#L103-L104

Reply:
Good spot with `user_directory_search`. I wasn't expecting that. Thankfully the no-op locking saves us.

But if locking on sqlite has never done anything, then the rationale for `lock=False` cannot have been because sqlite had no native upserts. Which means it must have been added as an optimization for the background update case on postgres?

Reply:
Yep, it's not unsafe, it just means this PR is rolling back a previous optimization we decided to add.